### DISCUSSION TITLE
How should an app with it's own update mechanism update the reported version in winget?

### DISCUSSION THREAD

**Discussion by josh-hemphill** [microsoft/winget-cli#3304] [Q&A]
The title kinda says it all.
Tried to find any mention of this and turned up nothing. I was wanting to submit a PR to pnpm, since it's almost always updated with other global JS packages, and wanted to clean up my winget upgrade list.

**Comment by mdanish-kh** [comment]
WinGet fetches the package metadata including version info from the registry. Most installers write the updated value to Registry in the auto-update scenario that informs WinGet of the updated value. Since pnpm is a `portable` (loose exe file) package and doesn't have an installer/uninstaller, WinGet still treats it as first-class package by making it available in PATH variable and handling the process of writing Registry keys for it. This makes it so that the package is visible in Apps&Features, and can be installed/upgraded/uninstalled through WinGet.

Correct me if I'm wrong (and you can verify this from `Get-Command "pnpm" | Select Path`) but what's likely happening is that you have another executable of pnpm in your machine that is available earlier in the PATH variable that's getting updated externally. Since it's just a standalone exe file present somewhere on your machine and the Registry isn't getting updated, WinGet has no way to detect its presence/version. If updates are being handled externally for a `portable` package, doesn't that defeat the purpose of using `winget upgrade` for that package? 

If you want to clean the upgrade list, you can either pin the package using `winget pin` so that WinGet doesn't attempt to upgrade it, or uninstall the package (this shouldn't cause issues if it's available through another directory)

**Reply by josh-hemphill** [reply]
> doesn't that defeat the purpose of using `winget upgrade` for that package?

No, I use winget to set up fresh installs with all my dev software with a single manifest file. After that, I'm usually working in a project with pnpm when I see there's a new version, and just tell it to update itself. I usually only go to winget for upgrades when there's some software that doesn't update itself I want to check on.
